File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& m) {
        vector<int> res;
  int left=0,right=m[0].size();
  int top=0,bottom=m.size();

  while(top<bottom && left<right)
    {
      for(int i=left;i<right;i++){
        res.push_back(m[top][i]);
      }
      top+=1;

      for(int i=top;i<bottom;i++){
        res.push_back(m[i][right-1]);
      }

      right-=1;

      if (!(top<bottom && left<right))
      {
        break;
      }

      for(int i=right-1;i>left-1;i--){
        res.push_back(m[bottom-1][i]);
      }

      bottom-=1;

      for(int i=bottom-1;i>top-1;i--){
        res.push_back(m[i][left]);
      }

      left+=1;
    }
        
        return res;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix:
            return []

        rows = len(matrix)
        cols = len(matrix[0])
        total_elements = rows * cols
        ans = []

        # Define the boundaries
        top = 0
        left = 0
        bottom = rows - 1
        right = cols - 1

        while len(ans) < total_elements:
            # Traverse the top row from left to right
            for i in range(left, right + 1):
                ans.append(matrix[top][i])
            top += 1

            # Traverse the right column from top to bottom
            for i in range(top, bottom + 1):
                ans.append(matrix[i][right])
            right -= 1

            # Check if there are remaining rows to traverse
            if len(ans) < total_elements and top <= bottom:
                # Traverse the bottom row from right to left
                for i in range(right, left - 1, -1):
                    ans.append(matrix[bottom][i])
                bottom -= 1

            # Check if there are remaining columns to traverse
            if len(ans) < total_elements and left <= right:
                # Traverse the left column from bottom to top
                for i in range(bottom, top - 1, -1):
                    ans.append(matrix[i][left])
                left += 1

        return ans